base/sort.jl

```diff
@@ -460,7 +460,51 @@ end
 
 ## sorting multi-dimensional arrays ##
 
-sort(A::AbstractArray, dim::Integer; kws...) = mapslices(a->sort(a; kws...), A, [dim])
+sort(A::AbstractArray, dim::Integer; kws...) = sortdim!(copy(A), dim; kws...)
+sort!(A::AbstractArray, dim::Integer; kws...) = sortdim!(A, dim; kws...)
+
+function sortdim!(A::Vector, dim::Integer; kws...)
+    dim != 1 && throw(ArgumentError("Bad dimension to sort."))
+    sort!(A; kws...)
+end
+
+function sortdim!(A::AbstractMatrix, dim::Integer; kws...)
+    if dim < 1 || dim > 2
+       throw(ArgumentError("Bad dimension to sort."))
+    end
+
+    if dim == 1
+        for idx = 1:size(A,2)
+            @inbounds sort!(slice(A, :, idx); kws...)
+        end
+    else
+        for idx = 1:size(A,1)
+            @inbounds sort!(slice(A, idx, :); kws...)
+        end
+    end
+
+    return A
+end
+
+function sortdim!(A::AbstractArray, dim::Integer; kws...)
+    ndimsA = ndims(A)
+    dimsA = size(A)
+
+    if dim < 1 || dim > ndimsA
+        throw(ArgumentError("Bad dimension to sort."))
+    end
+
+    pre_dims = dimsA[1:dim-1]
+    post_dims = dimsA[dim+1:end]
+
+    @inbounds for post_idx in CartesianRange(post_dims)
```

Comment:
The dimensionality of `CartesianRange(post_dims)` cannot be inferred, since `dim` is a value. It will be faster with a function barrier:
```
post_range = CartesianRange(post_dims)
pre_range = CartesianRange(pre_dims)
_sortdim!(A, pre_range, post_range; kws...)
```
and put `@noinline` in front of `_sortdim!`.

That said, it seems likely that most uses will be dominated by the actual `sort!` operation.